File: src/parser.ts
// Recursive-descent parser. Mirrors the TypeScript compiler parser: module-level
// mutable state initialized per parseSourceFile call, finishNode stamping
// pos/end and error flags, and list parsing (parseList/parseDelimitedList) with
// context-aware error recovery that always makes forward progress.
//
// This file holds the parser CORE (M3): scanning helpers, node construction,
// diagnostics, the list infrastructure, forEachChild, and a parseSourceFile that
// so far only recognizes EmptyStatement (";"). Grammar is layered on in later
// milestones; everything else is recovered as "declaration or statement
// expected" without aborting.

import { createScanner } from "./scanner.ts";
import { createDiagnostic } from "./diagnostics.ts";
import { Diagnostics } from "./diagnostics.ts";
import { tokenToString } from "./utilities.ts";
import {
	type Diagnostic,
	type DiagnosticMessage,
	type EmptyStatement,
	type Identifier,
	type Node,
	type NodeArray,
	NodeFlags,
	type Scanner,
	type SourceFile,
	type Statement,
	SyntaxKind,
	type Token,
} from "./types.ts";

type Mutable<T> = { -readonly [P in keyof T]: T[P] };

const enum ParsingContext {
	SourceElements,
	BlockStatements,
	ClassMembers,
	EnumConstants,
	TypeArguments,
	TypeParameters,
	Parameters,
	ArgumentExpressions,
	VariableDeclarations,
	ArrayInitializerElements,
	SwitchClauses,
	CatchClauses,
	Count,
}

// Module-level parser state, reset at the start of every parseSourceFile call.
let scanner: Scanner;
let sourceText = "";
let fileName = "";
let currentToken: SyntaxKind = SyntaxKind.Unknown;
let parseDiagnostics: Diagnostic[] = [];
let parsingContext = 0;
let parseErrorBeforeNextFinishedNode = false;

function token(): SyntaxKind {
	return currentToken;
}

function nextToken(): SyntaxKind {
	return (currentToken = scanner.scan());
}

function getNodePos(): number {
	return scanner.getTokenFullStart();
}

function createNode<T extends Node>(kind: SyntaxKind, pos: number): Mutable<T> {
	return {
		kind,
		flags: NodeFlags.None,
		pos,
		end: pos,
		parent: undefined,
	} as unknown as Mutable<T>;
}

function finishNode<T extends Node>(node: Mutable<T>, pos: number, end = scanner.getTokenFullStart()): T {
	node.pos = pos;
	node.end = end;
	if (parseErrorBeforeNextFinishedNode) {
		parseErrorBeforeNextFinishedNode = false;
		node.flags |= NodeFlags.ThisNodeHasError;
	}
	return node as T;
}

function createNodeArray<T extends Node>(elements: T[], pos: number, end = getNodePos()): NodeArray<T> {
	const array = elements as unknown as Mutable<NodeArray<T>>;
	array.pos = pos;
	array.end = end;
	return array as NodeArray<T>;
}

// Diagnostics

function parseErrorAtPosition(
	start: number,
	length: number,
	message: DiagnosticMessage,
	...args: string[]
): void {
	const last = parseDiagnostics[parseDiagnostics.length - 1];
	if (!last || start !== last.pos) {
		parseDiagnostics.push(createDiagnostic(start, length, message, ...args));
	}
	// Tell the next finishNode that the node it completes spans a parse error.
	parseErrorBeforeNextFinishedNode = true;
}

function parseErrorAt(start: number, end: number, message: DiagnosticMessage, ...args: string[]): void {
	parseErrorAtPosition(start, end - start, message, ...args);
}

function parseErrorAtCurrentToken(message: DiagnosticMessage, ...args: string[]): void {
	parseErrorAt(scanner.getTokenStart(), scanner.getTokenEnd(), message, ...args);
}

// Token consumption helpers

function parseExpected(kind: SyntaxKind, message?: DiagnosticMessage): boolean {
	if (token() === kind) {
		nextToken();
		return true;
	}
	if (message) {
		parseErrorAtCurrentToken(message);
	} else {
		parseErrorAtCurrentToken(Diagnostics._0_expected, tokenToString(kind) ?? "");
	}
	return false;
}

function parseOptional(kind: SyntaxKind): boolean {
	if (token() === kind) {
		nextToken();
		return true;
	}
	return false;
}

function parseTokenNode<T extends Node>(): T {
	const pos = getNodePos();
	const kind = token();
	nextToken();
	return finishNode(createNode<T>(kind, pos), pos);
}

function parseOptionalToken<T extends Node>(kind: SyntaxKind): T | undefined {
	if (token() === kind) {
		return parseTokenNode<T>();
	}
	return undefined;
}

function parseExpectedToken<T extends Node>(kind: SyntaxKind, message?: DiagnosticMessage): T {
	return (
		parseOptionalToken<T>(kind) ??
		createMissingNode<T>(kind, /*reportAtCurrentPosition*/ false, message ?? Diagnostics._0_expected, tokenToString(kind) ?? "")
	);
}

function createMissingNode<T extends Node>(
	kind: SyntaxKind,
	reportAtCurrentPosition: boolean,
	message?: DiagnosticMessage,
	...args: string[]
): T {
	if (reportAtCurrentPosition) {
		parseErrorAtPosition(scanner.getTokenFullStart(), 0, message ?? Diagnostics._0_expected, ...args);
	} else if (message) {
		parseErrorAtCurrentToken(message, ...args);
	}
	const pos = getNodePos();
	const node = createNode<T>(kind, pos);
	if (kind === SyntaxKind.Identifier) {
		(node as unknown as Mutable<Identifier>).text = "";
	}
	return finishNode(node, pos);
}

function parseIdentifier(): Identifier {
	if (token() === SyntaxKind.Identifier) {
		const pos = getNodePos();
		const text = scanner.getTokenValue();
		nextToken();
		const node = createNode<Identifier>(SyntaxKind.Identifier, pos);
		node.text = text;
		return finishNode(node, pos);
	}
	return createMissingNode<Identifier>(SyntaxKind.Identifier, /*reportAtCurrentPosition*/ false, Diagnostics.Identifier_expected);
}

// List parsing with error recovery

function isListTerminator(context: ParsingContext): boolean {
	if (token() === SyntaxKind.EndOfFileToken) {
		return true;
	}
	switch (context) {
		case ParsingContext.BlockStatements:
		case ParsingContext.ClassMembers:
		case ParsingContext.EnumConstants:
		case ParsingContext.SwitchClauses:
		case ParsingContext.ArrayInitializerElements:
			return token() === SyntaxKind.CloseBraceToken;
		case ParsingContext.Parameters:
		case ParsingContext.ArgumentExpressions:
			return token() === SyntaxKind.CloseParenToken;
		case ParsingContext.TypeArguments:
		case ParsingContext.TypeParameters:
			return token() === SyntaxKind.GreaterThanToken;
		default:
			return false;
	}
}

function isStartOfStatement(): boolean {
	// M3 stub: only the empty statement is recognized. Extended in M7.
	return token() === SyntaxKind.SemicolonToken;
}

function isListElement(context: ParsingContext, _inErrorRecovery: boolean): boolean {
	switch (context) {
		case ParsingContext.SourceElements:
		case ParsingContext.BlockStatements:
			return isStartOfStatement();
		default:
			return false;
	}
}

function isInSomeParsingContext(): boolean {
	for (let context = 0; context < ParsingContext.Count; context++) {
		if (parsingContext & (1 << context)) {
			if (isListElement(context, /*inErrorRecovery*/ true) || isListTerminator(context)) {
				return true;
			}
		}
	}
	return false;
}

function parsingContextErrors(context: ParsingContext): void {
	switch (context) {
		case ParsingContext.SourceElements:
		case ParsingContext.BlockStatements:
			parseErrorAtCurrentToken(Diagnostics.Declaration_or_statement_expected);
			return;
		case ParsingContext.Parameters:
			parseErrorAtCurrentToken(Diagnostics.Parameter_declaration_expected);
			return;
		default:
			parseErrorAtCurrentToken(Diagnostics.Unexpected_token);
	}
}

// Decide whether to abort the current list (because the token is valid in some
// enclosing context) or skip the offending token and keep going. Always makes
// progress, so list parsing cannot loop forever.
function abortParsingListOrMoveToNextToken(context: ParsingContext): boolean {
	parsingContextErrors(context);
	if (isInSomeParsingContext()) {
		return true;
	}
	nextToken();
	return false;
}

function parseList<T extends Node>(context: ParsingContext, parseElement: () => T): NodeArray<T> {
	const saveParsingContext = parsingContext;
	parsingContext |= 1 << context;
	const list: T[] = [];
	const listPos = getNodePos();

	while (!isListTerminator(context)) {
		if (isListElement(context, /*inErrorRecovery*/ false)) {
			list.push(parseElement());
			continue;
		}
		if (abortParsingListOrMoveToNextToken(context)) {
			break;
		}
	}

	parsingContext = saveParsingContext;
	return createNodeArray(list, listPos);
}

// Statements (M3: empty statement only)

function parseEmptyStatement(): EmptyStatement {
	const pos = getNodePos();
	parseExpected(SyntaxKind.SemicolonToken);
	return finishNode(createNode<EmptyStatement>(SyntaxKind.EmptyStatement, pos), pos);
}

function parseStatement(): Statement {
	// M3 stub matching isStartOfStatement.
	return parseEmptyStatement();
}

// Entry point

export function parseSourceFile(fileNameArg: string, text: string): SourceFile {
	fileName = fileNameArg;
	sourceText = text;
	parseDiagnostics = [];
	parsingContext = 0;
	parseErrorBeforeNextFinishedNode = false;
	scanner = createScanner(text, (message, errPos, length) => parseErrorAtPosition(errPos, length, message));

	nextToken();
	const pos = getNodePos();
	const statements = parseList(ParsingContext.SourceElements, parseStatement);
	const endOfFileToken = parseExpectedToken<Token<SyntaxKind.EndOfFileToken>>(SyntaxKind.EndOfFileToken);

	const node = createNode<SourceFile>(SyntaxKind.SourceFile, pos);
	node.statements = statements;
	node.endOfFileToken = endOfFileToken;
	node.fileName = fileName;
	node.text = sourceText;
	finishNode(node, pos);
	node.parseDiagnostics = parseDiagnostics;
	return node;
}

// Tree walking. Mirrors the TS compiler forEachChild: visit each child node (and
// child NodeArray) in source order, returning the first truthy callback result.
// The binder and every LSP traversal go through this.

function visitNode<T>(cbNode: (node: Node) => T | undefined, node: Node | undefined): T | undefined {
	return node ? cbNode(node) : undefined;
}

function visitNodes<T>(
	cbNode: (node: Node) => T | undefined,
	cbNodes: ((nodes: NodeArray<Node>) => T | undefined) | undefined,
	nodes: NodeArray<Node> | undefined,
): T | undefined {
	if (!nodes) return undefined;
	if (cbNodes) return cbNodes(nodes);
	for (const node of nodes) {
		const result = cbNode(node);
		if (result) return result;
	}
	return undefined;
}

export function forEachChild<T>(
	node: Node,
	cbNode: (node: Node) => T | undefined,
	cbNodes?: (nodes: NodeArray<Node>) => T | undefined,
): T | undefined {
	switch (node.kind) {
		case SyntaxKind.SourceFile: {
			const sf = node as SourceFile;
			return visitNodes(cbNode, cbNodes, sf.statements) || visitNode(cbNode, sf.endOfFileToken);
		}
		default:
			// Tokens and childless nodes (EmptyStatement, Identifier, ...).
			return undefined;
	}
}

File: togo/internal/compiler/parser.go
package compiler

// Recursive-descent parser core. The TS original uses module-level mutable
// state; following tsgo this becomes a Parser struct with methods. finishNode
// stamps pos/end and the error flag; parseList/parseDelimitedList do
// context-aware error recovery that always makes forward progress. Port of the
// parser core in src/compiler/parser.ts.
//
// Grammar coverage grows in slices: this step is the compilation-unit skeleton
// (empty statements + recovery + EOF). isStartOfStatement is scoped to ';'
// until the type-declaration grammar is ported.

// ParsingContext identifies the list currently being parsed (for terminators
// and recovery). Values mirror the TS enum.
type ParsingContext int

const (
	ctxSourceElements ParsingContext = iota
	ctxBlockStatements
	ctxClassMembers
	ctxEnumConstants
	ctxTypeArguments
	ctxTypeParameters
	ctxParameters
	ctxArgumentExpressions
	ctxVariableDeclarations
	ctxArrayInitializerElements
	ctxSwitchClauses
	ctxCatchClauses
	ctxModuleDirectives
	ctxAnnotationValues
	ctxCount
)

// Parser holds the per-parse mutable state.
type Parser struct {
	scanner                          *Scanner
	factory                          NodeFactory
	sourceText                       string
	fileName                         string
	currentToken                     SyntaxKind
	parseDiagnostics                 []Diagnostic
	parsingContext                   int
	parseErrorBeforeNextFinishedNode bool
}

// ParseSourceFile parses text into a SourceFile node.
func ParseSourceFile(fileName, text string) *Node {
	p := &Parser{sourceText: text, fileName: fileName, currentToken: Unknown}
	p.scanner = NewScanner(text, func(message DiagnosticMessage, errPos, length int) {
		p.parseErrorAtPosition(errPos, length, message)
	})

	p.nextToken()
	pos := p.getNodePos()
	var pkg *Node
	if p.token() == PackageKeyword {
		pkg = p.parsePackageDeclaration()
	}
	imports := p.parseImportDeclarations()
	// module declarations and type declarations arrive with later grammar slices.
	statements := p.parseList(ctxSourceElements, p.parseSourceElement)
	eof := p.parseExpectedToken(EndOfFileToken, nil)

	sf := p.factory.NewSourceFile(pkg, imports, statements, eof, nil)
	p.finishNode(sf, pos, -1)
	data := sf.AsSourceFile()
	data.FileName = p.fileName
	data.Text = p.sourceText
	data.ParseDiagnostics = p.parseDiagnostics
	return sf
}

func (p *Parser) token() SyntaxKind     { return p.currentToken }
func (p *Parser) nextToken() SyntaxKind { p.currentToken = p.scanner.Scan(); return p.currentToken }
func (p *Parser) getNodePos() int       { return p.scanner.TokenFullStart() }

// finishNode stamps pos/end (end<0 means "current full start") and the error flag.
func (p *Parser) finishNode(node *Node, pos, end int) *Node {
	if end < 0 {
		end = p.scanner.TokenFullStart()
	}
	node.Pos = pos
	node.End = end
	if p.parseErrorBeforeNextFinishedNode {
		p.parseErrorBeforeNextFinishedNode = false
		node.Flags |= NodeFlagThisNodeHasError
	}
	return node
}

func (p *Parser) createNodeArray(nodes []*Node, pos, end int) *NodeArray {
	if end < 0 {
		end = p.getNodePos()
	}
	return &NodeArray{Nodes: nodes, Pos: pos, End: end}
}

// --- diagnostics -------------------------------------------------------------

func (p *Parser) parseErrorAtPosition(start, length int, message DiagnosticMessage, args ...string) {
	if n := len(p.parseDiagnostics); n == 0 || start != p.parseDiagnostics[n-1].Pos {
		p.parseDiagnostics = append(p.parseDiagnostics, CreateDiagnostic(start, length, message, args...))
	}
	// Tell the next finishNode that the node it completes spans a parse error.
	p.parseErrorBeforeNextFinishedNode = true
}

func (p *Parser) parseErrorAt(start, end int, message DiagnosticMessage, args ...string) {
	p.parseErrorAtPosition(start, end-start, message, args...)
}

func (p *Parser) parseErrorAtCurrentToken(message DiagnosticMessage, args ...string) {
	p.parseErrorAt(p.scanner.TokenStart(), p.scanner.TokenEnd(), message, args...)
}

// --- token consumption -------------------------------------------------------

func (p *Parser) parseExpected(kind SyntaxKind, message *DiagnosticMessage) bool {
	if p.token() == kind {
		p.nextToken()
		return true
	}
	if message != nil {
		p.parseErrorAtCurrentToken(*message)
	} else {
		p.parseErrorAtCurrentToken(Diagnostics.Expected0, tokenToString(kind))
	}
	return false
}

func (p *Parser) parseOptional(kind SyntaxKind) bool {
	if p.token() == kind {
		p.nextToken()
		return true
	}
	return false
}

func (p *Parser) parseTokenNode() *Node {
	pos := p.getNodePos()
	kind := p.token()
	p.nextToken()
	return p.finishNode(p.factory.newToken(kind), pos, -1)
}

func (p *Parser) parseExpectedToken(kind SyntaxKind, message *DiagnosticMessage) *Node {
	if p.token() == kind {
		return p.parseTokenNode()
	}
	return p.createMissingNode(kind, false, message, tokenToString(kind))
}

func (p *Parser) createMissingNode(kind SyntaxKind, reportAtCurrentPosition bool, message *DiagnosticMessage, args ...string) *Node {
	switch {
	case reportAtCurrentPosition:
		m := Diagnostics.Expected0
		if message != nil {
			m = *message
		}
		p.parseErrorAtPosition(p.scanner.TokenFullStart(), 0, m, args...)
	case message != nil:
		p.parseErrorAtCurrentToken(*message, args...)
	}
	pos := p.getNodePos()
	var node *Node
	if kind == Identifier {
		node = p.factory.NewIdentifier("")
	} else {
		node = p.factory.newToken(kind)
	}
	return p.finishNode(node, pos, -1)
}

// parserLookAhead peeks with cb without consuming input: it saves and restores
// the parser-level state (current token, diagnostics, error flag) on top of the
// scanner's lookahead. Mirrors speculationHelper(callback, isLookahead=true).
func parserLookAhead[T any](p *Parser, cb func() T) T {
	saveToken := p.currentToken
	saveLen := len(p.parseDiagnostics)
	saveErr := p.parseErrorBeforeNextFinishedNode
	result := LookAhead(p.scanner, cb)
	p.currentToken = saveToken
	p.parseDiagnostics = p.parseDiagnostics[:saveLen]
	p.parseErrorBeforeNextFinishedNode = saveErr
	return result
}

func (p *Parser) parseIdentifier() *Node {
	if p.token() == Identifier {
		pos := p.getNodePos()
		text := p.scanner.TokenValue()
		p.nextToken()
		return p.finishNode(p.factory.NewIdentifier(text), pos, -1)
	}
	return p.createMissingNode(Identifier, false, &Diagnostics.IdentifierExpected)
}

// --- list parsing with error recovery ----------------------------------------

func (p *Parser) isListTerminator(context ParsingContext) bool {
	if p.token() == EndOfFileToken {
		return true
	}
	switch context {
	case ctxBlockStatements, ctxClassMembers, ctxEnumConstants, ctxSwitchClauses,
		ctxArrayInitializerElements, ctxModuleDirectives:
		return p.token() == CloseBraceToken
	case ctxParameters, ctxArgumentExpressions, ctxAnnotationValues:
		return p.token() == CloseParenToken
	case ctxTypeArguments, ctxTypeParameters:
		return p.token() == GreaterThanToken
	default:
		return false
	}
}

func (p *Parser) isListElement(context ParsingContext) bool {
	switch context {
	case ctxSourceElements:
		return p.isStartOfStatement()
	case ctxClassMembers:
		return p.isStartOfClassMember()
	case ctxTypeArguments:
		return p.isStartOfType()
	case ctxTypeParameters:
		return p.token() == Identifier || p.token() == AtToken
	case ctxParameters:
		return p.isStartOfParameter()
	case ctxArgumentExpressions:
		return p.isStartOfExpression()
	default:
		return false
	}
}

func (p *Parser) isInSomeParsingContext() bool {
	for context := ParsingContext(0); context < ctxCount; context++ {
		if p.parsingContext&(1<<context) != 0 {
			if p.isListElement(context) || p.isListTerminator(context) {
				return true
			}
		}
	}
	return false
}

func (p *Parser) parsingContextErrors(context ParsingContext) {
	switch context {
	case ctxSourceElements, ctxBlockStatements:
		p.parseErrorAtCurrentToken(Diagnostics.DeclarationOrStatementExpected)
	case ctxParameters:
		p.parseErrorAtCurrentToken(Diagnostics.ParameterDeclarationExpected)
	default:
		p.parseErrorAtCurrentToken(Diagnostics.UnexpectedToken)
	}
}

// abortParsingListOrMoveToNextToken aborts the current list when the token is
// valid in an enclosing context, else skips it. Always makes progress.
func (p *Parser) abortParsingListOrMoveToNextToken(context ParsingContext) bool {
	p.parsingContextErrors(context)
	if p.isInSomeParsingContext() {
		return true
	}
	p.nextToken()
	return false
}

func (p *Parser) parseList(context ParsingContext, parseElement func() *Node) *NodeArray {
	saveParsingContext := p.parsingContext
	p.parsingContext |= 1 << context
	var list []*Node
	listPos := p.getNodePos()

	for !p.isListTerminator(context) {
		if p.isListElement(context) {
			list = append(list, parseElement())
			continue
		}
		if p.abortParsingListOrMoveToNextToken(context) {
			break
		}
	}

	p.parsingContext = saveParsingContext
	return p.createNodeArray(list, listPos, -1)
}

// parseDelimitedList parses a comma-separated list, with the "skip a token if no
// progress" guard so malformed input cannot loop. Port of parseDelimitedList.
func (p *Parser) parseDelimitedList(context ParsingContext, parseElement func() *Node) *NodeArray {
	saveParsingContext := p.parsingContext
	p.parsingContext |= 1 << context
	var list []*Node
	listPos := p.getNodePos()

	for {
		if p.isListElement(context) {
			startPos := p.scanner.TokenFullStart()
			list = append(list, parseElement())
			if p.parseOptional(CommaToken) {
				continue
			}
			if p.isListTerminator(context) {
				break
			}
			p.parseExpected(CommaToken, nil)
			if startPos == p.scanner.TokenFullStart() {
				p.nextToken()
			}
			continue
		}
		if p.isListTerminator(context) {
			break
		}
		if p.abortParsingListOrMoveToNextToken(context) {
			break
		}
	}

	p.parsingContext = saveParsingContext
	return p.createNodeArray(list, listPos, -1)
}

// --- names -------------------------------------------------------------------

func (p *Parser) makeQualifiedName(left, right *Node) *Node {
	return p.finishNode(p.factory.NewQualifiedName(left, right), left.Pos, -1)
}

func (p *Parser) parseEntityName() *Node {
	entity := p.parseIdentifier()
	// Only consume a dot followed by an identifier, so a ".<T>" generic call is
	// left for the expression parser rather than eaten here.
	for p.token() == DotToken && parserLookAhead(p, func() bool { p.nextToken(); return p.token() == Identifier }) {
		p.parseExpected(DotToken, nil)
		entity = p.makeQualifiedName(entity, p.parseIdentifier())
	}
	return entity
}

// --- compilation unit --------------------------------------------------------

func (p *Parser) parsePackageDeclaration() *Node {
	pos := p.getNodePos()
	p.parseExpected(PackageKeyword, nil)
	name := p.parseEntityName()
	p.parseExpected(SemicolonToken, nil)
	return p.finishNode(p.factory.NewPackageDeclaration(nil, name), pos, -1)
}

func (p *Parser) parseImportDeclaration() *Node {
	pos := p.getNodePos()
	p.parseExpected(ImportKeyword, nil)
	isStatic := p.parseOptional(StaticKeyword)
	name := p.parseIdentifier()
	isOnDemand := false
	for p.parseOptional(DotToken) {
		if p.token() == AsteriskToken {
			p.nextToken()
			isOnDemand = true
			break
		}
		name = p.makeQualifiedName(name, p.parseIdentifier())
	}
	p.parseExpected(SemicolonToken, nil)
	return p.finishNode(p.factory.NewImportDeclaration(isStatic, name, isOnDemand), pos, -1)
}

func (p *Parser) parseImportDeclarations() *NodeArray {
	pos := p.getNodePos()
	var list []*Node
	for p.token() == ImportKeyword {
		list = append(list, p.parseImportDeclaration())
	}
	return p.createNodeArray(list, pos, -1)
}

// --- statements --------------------------------------------------------------

func (p *Parser) parseSourceElement() *Node {
	if p.token() == SemicolonToken {
		return p.parseEmptyStatement()
	}
	return p.parseTypeDeclaration()
}

func (p *Parser) parseEmptyStatement() *Node {
	pos := p.getNodePos()
	p.parseExpected(SemicolonToken, nil)
	return p.finishNode(p.factory.NewEmptyStatement(), pos, -1)
}
